# Statistical Methods in Pandas - Lab

## Introduction

In this lab you'll get some hands-on experience using some of the key summary statistics methods in Pandas.

## Objectives
You will be able to:

- Calculate summary statistics for a series and DataFrame
- Use the `.apply()` or `.applymap()` methods to apply a function to a pandas series or DataFrame  


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets. You will find this dataset in the file `'lego_sets.csv'` which is under the DATA folder.   

In the cell below:

- Import Pandas and set the standard alias of `pd`
- Import the `'lego_sets.csv'` dataset
- Display the first five rows of the DataFrame to get a feel for what we'll be working with

In [4]:
# Import pandas
import pandas as pd

from google.colab import drive
drive.mount('/content/drive')



with open("/content/drive/MyDrive/DATA/lego_sets.csv") as f:
    reader = csv.DictReader(f)
    lego_sets = list(reader)

# Import the 'lego_sets.csv' dataset
df = pd.read_csv('/content/drive/MyDrive/DATA/lego_sets.csv')

# Print the first five rows of DataFrame
for index in range(5):
    print(lego_sets[index])

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
{'ages': '6-12', 'list_price': '29.99', 'num_reviews': '2.0', 'piece_count': '277.0', 'play_star_rating': '4.0', 'prod_desc': 'Catapult into action and take back the eggs from the Piggy Trike!', 'prod_id': '75823.0', 'prod_long_desc': "Use the staircase catapult to launch Red into the air and race after the piggy, who is speeding off in the Piggy Trike with the eggs. Help him take the eggs out of the trike's box in this hilarious set, which also features Matilda's House to help Red relax, and a spinning Billy the Sign, based on The Angry BirdsTM Movie. Includes 3 figures.\nIncludes 3 figures: Red, Matilda and Biker Pig.\nBird Island Egg Heist features Matilda's House with a staircase catapult and easel, plus a branch, plants, flower elements and a Piggy Trike.\nPiggy Trike features a giant front wheel, 2 balloons and a box at the rear for eggs.\nAccessory ele

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset. There are two ways we'll get this information -- `.info()` and `.describe()`.

The `.info()` method provides us metadata on the DataFrame itself. This allows us to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset?
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method.

In [5]:
# Call the .info() method
df.info

<bound method DataFrame.info of        ages  list_price  num_reviews  piece_count  play_star_rating  \
0      6-12     29.9900          2.0        277.0               4.0   
1      6-12     19.9900          2.0        168.0               4.0   
2      6-12     12.9900         11.0         74.0               4.3   
3       12+     99.9900         23.0       1032.0               3.6   
4       12+     79.9900         14.0        744.0               3.2   
...     ...         ...          ...          ...               ...   
12256  7-14     36.5878          6.0        341.0               4.4   
12257  7-14     24.3878          8.0        217.0               4.1   
12258  7-14     24.3878         18.0        233.0               4.6   
12259  6-14     12.1878          1.0         48.0               5.0   
12260  6-14     12.1878         11.0        109.0               4.5   

                                               prod_desc  prod_id  \
0      Catapult into action and take back the eggs fr...  75823.0   
1      Launch a flying attack and rescue the eggs fro...  75822.0   
2      Chase the piggy with lightning-fast Chuck and ...  75821.0   
3      Explore the architecture of the United States ...  21030.0   
4      Recreate the Solomon R. Guggenheim Museum® wit...  21035.0   
...                                                  ...      ...   
12256  Protect NINJAGO® City from flying Manta Ray Bo...  70609.0   
12257         Stop a Piranha Attack with Kai and Misako!  70629.0   
12258   Stop a crime in the NINJAGO® City street market!  70607.0   
12259  Achieve Spinjitzu greatness with the Green Ninja!  70628.0   
12260  Practice your Spinjitzu skills with Kai and Zane!  70606.0   

                                          prod_long_desc review_difficulty  \
0      Use the staircase catapult to launch Red into ...           Average   
1      Pilot Pig has taken off from Bird Island with ...              Easy   
2      Pitch speedy bird Chuck against the Piggy Car....              Easy   
3      Discover the architectural secrets of the icon...           Average   
4      Discover the architectural secrets of Frank Ll...       Challenging   
...                                                  ...               ...   
12256  Help Cole save Shen-Li in this cool THE LEGO® ...              Easy   
12257  Play out an action-packed Piranha Mech pursuit...              Easy   
12258  Team up with Lloyd Garmadon, Nya and Officer T...              Easy   
12259  Learn all the skills of Spinjitzu with THE LEG...         Very Easy   
12260  Join the ninja heroes at the dojo with this ac...              Easy   

                             set_name  star_rating                 theme_name  \
0               Bird Island Egg Heist          4.5               Angry Birds™   
1                  Piggy Plane Attack          5.0               Angry Birds™   
2                    Piggy Car Escape          4.3               Angry Birds™   
3      United States Capitol Building          4.6               Architecture   
4       Solomon R. Guggenheim Museum®          4.6               Architecture   
...                               ...          ...                        ...   
12256                Manta Ray Bomber          4.3  THE LEGO® NINJAGO® MOVIE™   
12257                  Piranha Attack          3.6  THE LEGO® NINJAGO® MOVIE™   
12258             NINJAGO® City Chase          4.6  THE LEGO® NINJAGO® MOVIE™   
12259        Lloyd - Spinjitzu Master          5.0  THE LEGO® NINJAGO® MOVIE™   
12260              Spinjitzu Training          4.7  THE LEGO® NINJAGO® MOVIE™   

       val_star_rating country  
0                  4.0      US  
1                  4.0      US  
2                  4.1      US  
3                  4.3      US  
4                  4.1      US  
...                ...     ...  
12256              4.2      PT  
12257              4.1      PT  
12258              4.5      PT  
12259              5.0      PT  
12260        

#### Interpreting the Results

Read the output above, and then answer the following questions:

- How many total rows are in this DataFrame?   
- How many columns contain numeric data?
- How many contain categorical data?
- Identify at least 3 columns that contain missing values.

Write your answer below this line in comment form:
___________________________________________________

*   12,261
*   7
*   7
*   review_difficulty has 2055 missing values
_____________________________________________________________________________

In [14]:
#How many total rows are in this DataFrame?
print(len(df))

#How many columns contain numeric data?
print(df.select_dtypes(include='number').shape[1])

#How many contain categorical data?
print(df.select_dtypes(include='object').shape[1])

#Identify at least 3 columns that contain missing values.
print(df.isnull().sum().sort_values(ascending=False).head(10))


12261
7
7
review_difficulty    2055
val_star_rating      1795
play_star_rating     1775
star_rating          1620
num_reviews          1620
prod_desc             377
theme_name              3
piece_count             0
ages                    0
list_price              0
dtype: int64


Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method.

In [15]:
# Call the .describe() method
df.describe()

list_price   num_reviews   piece_count  play_star_rating  \
count  12261.000000  10641.000000  12261.000000      10486.000000   
mean      65.141998     16.826238    493.405921          4.337641   
std       91.980429     36.368984    825.364580          0.652051   
min        2.272400      1.000000      1.000000          1.000000   
25%       19.990000      2.000000     97.000000          4.000000   
50%       36.587800      6.000000    216.000000          4.500000   
75%       70.192200     13.000000    544.000000          4.800000   
max     1104.870000    367.000000   7541.000000          5.000000   

            prod_id   star_rating  val_star_rating  
count  1.226100e+04  10641.000000     10466.000000  
mean   5.983675e+04      4.514134         4.228960  
std    1.638115e+05      0.518865         0.660282  
min    6.300000e+02      1.800000         1.000000  
25%    2.103400e+04      4.300000         4.000000  
50%    4.206900e+04      4.700000         4.300000  
75%    7.092200e+04      5.000000         4.700000  
max    2.000431e+06      5.000000         5.000000

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns. Use these to answer the following questions:

- How much is the standard deviation for `piece count`?
- How many pieces are in the largest lego set?
- How many in the smallest lego set? What is the median `val_star_rating`?

___________________________________________________

*   825.364580
*   List item

_____________________________________________________________________________

In [20]:

#please use this space to answer the questions
std_piece_count = df['piece_count'].std()
max_piece_count = df['piece_count'].max()
min_piece_count = df['piece_count'].min()

print(f"The standard deviation for piece count is {std_piece_count}")
print(f"The largest lego set has {max_piece_count} pieces")
print(f"The smallest lego set has {min_piece_count} pieces")

The standard deviation for piece count is 825.364580411521
The largest lego set has 7541.0 pieces
The smallest lego set has 1.0 pieces


## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods.

In the cell below, compute the median value of the `star_rating` column.

In [21]:
# Calculate the median of the star_rating column (Hint: there is a median function)
median_star_rating = df['star_rating'].median()
print(median_star_rating)

4.7


Next, get a count of the total number of unique values in `play_star_rating`.

In [22]:
# Print the number of unique values in play_star_rating
unique_play_star_rating = df['play_star_rating'].nunique()
print(unique_play_star_rating)

30


Now, compute the standard deviation of the `list_price` column.

In [24]:
# Calculate the standard deviation of the list_price column. (Hint: you can use std function)
std_list_price = df['list_price'].std()
print(std_list_price)

91.9804293059252


If we bought every single lego set in this dataset, how many pieces would we have?  

> **Note**: If you truly want to answer this accurately, and are up for the challenge, remove duplicate lego-set entries before summing the pieces. That is, many of the lego sets are listed multiple times in the dataset above, depending on the country where it is being sold and other unique parameters. If you're stuck, just practice calculating the total number of pieces in the dataset for now.

In [26]:
# Total number of pieces across all unique Lego sets
df.drop_duplicates(subset='prod_id')['piece_count'].sum()

# If you're simply calculating the sum of a column
# df['piece_count'].sum()



np.float64(319071.0)

Now, let's try getting the value for the 90% quantile for all numerical columns.  Do this in the cell below.

In [31]:
# Get the 90% quantile for all numerical columns. (Hint: quantile function)
df.select_dtypes(include=['number']).quantile([0.9])

list_price  num_reviews  piece_count  play_star_rating  prod_id  \
0.9    136.2971         38.0       1077.0               5.0  75531.0   

     star_rating  val_star_rating  
0.9          5.0              5.0

## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data -- there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data.

In the cell below, print the unique values contained within the `review_difficulty` column.

In [32]:
# Print the unique values in the review_difficulty column
df['review_difficulty'].unique()

array(['Average', 'Easy', 'Challenging', 'Very Easy', nan,
       'Very Challenging'], dtype=object)

Now, let's get the `value_counts()` for this column, to see how common each is.

In [33]:
# Get the value_counts() of the review_difficulty column
print(df['review_difficulty'].value_counts())

# Alternatively normalized value counts
df['review_difficulty'].value_counts(normalize=True)

review_difficulty
Easy                4236
Average             3765
Very Easy           1139
Challenging         1058
Very Challenging       8
Name: count, dtype: int64


review_difficulty
Easy                0.415050
Average             0.368901
Very Easy           0.111601
Challenging         0.103665
Very Challenging    0.000784
Name: proportion, dtype: float64


## Using `.groupby`

For each review difficulty level, find how many legos you have. Then sort these based on their counts. (Hint: Please search for groupby in Pandas, `size()`, `count()`, `reset_index()`, and `sort_values()` functions)



As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `.applymap()` method and passing in a lambda function.

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call the DataFrame's `.applymap()` method and pass in `lambda x: str(x)`  
* Call the new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [36]:
# Call the .applymap() method
string_df = df.map(lambda x: str(x))

# Call the .info() method
string_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ages               12261 non-null  object
 1   list_price         12261 non-null  object
 2   num_reviews        12261 non-null  object
 3   piece_count        12261 non-null  object
 4   play_star_rating   12261 non-null  object
 5   prod_desc          12261 non-null  object
 6   prod_id            12261 non-null  object
 7   prod_long_desc     12261 non-null  object
 8   review_difficulty  12261 non-null  object
 9   set_name           12261 non-null  object
 10  star_rating        12261 non-null  object
 11  theme_name         12261 non-null  object
 12  val_star_rating    12261 non-null  object
 13  country            12261 non-null  object
dtypes: object(14)
memory usage: 1.3+ MB


/tmp/ipython-input-1211924104.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_df = df.applymap(lambda x: str(x))


Note that everything -- even the `NaN` values, have been cast to a string in the example above.

Note that for Pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `.apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame

## Summary

In this lab, we learned how to:

* Use the `df.describe()` and `df.info()` summary statistics methods
* Use built-in Pandas methods for calculating summary statistics
* Apply a function to every element in a DataFrame